resteasy-wadl/src/main/java/org/jboss/resteasy/wadl/ResteasyWadlWriter.java

```diff
@@ -232,4 +283,188 @@ private void setType(Param param, ResteasyWadlMethodParamMetaData paramMetaData)
+               } else {
+                  LogMessages.LOGGER.error(Messages.MESSAGES.cantProcessWadl());
+               }
+           } catch (Exception e) {
+              LogMessages.LOGGER.error(Messages.MESSAGES.cantProcessWadl());
+           }
+       }
+
+       public void enableSchemaGeneration() {
+           generateSchema = true;
+       }
+
+       public boolean schemaGenerationEnabled() {
+           return generateSchema;
+       }
+
+       protected void addGrammars(Grammars grammars) {
+           if (this.grammars == null) {
+               this.grammars = grammars;
+           } else {
+               if (!grammars.getAny().isEmpty()) {
+                   this.grammars.getAny().addAll(grammars.getAny());
+               }
+               if (!grammars.getDoc().isEmpty()) {
+                   this.grammars.getDoc().addAll(grammars.getDoc());
+               }
+               if (!grammars.getInclude().isEmpty()) {
+                   this.grammars.getInclude().addAll(grammars.getInclude());
+               }
+           }
+       }
+
+       private void collectClassesForSchemaGeneration(ResteasyWadlMethodMetaData methodMetaData) {
+           if (!schemaGenerationEnabled())
+               return;
+           // support runtime rescan.
+           schemaClasses.clear();
+           generatedSchemas.clear();
+
+           _addClass(methodMetaData.getMethod().getReturnType());
+
+           for (ResteasyWadlMethodParamMetaData paramMetaData : methodMetaData.getParameters()) {
+               _addClass(paramMetaData.getType());
+           }
+
+           processClassesForSchema();
+       }
+
+       private void _addClass(Class clazz) {
+           if (clazz.getAnnotation(XmlRootElement.class) != null) {
+               schemaClasses.add(clazz);
+           }
+       }
+
+       private void processClassesForSchema() {
+           try {
+               final JAXBContext context = JAXBContext.newInstance(schemaClasses.toArray(new Class[schemaClasses.size()]));
+
+               final List<StreamResult> results = new ArrayList<>();
+
+               context.generateSchema(new SchemaOutputResolver() {
+                   int counter = 0;
+
+                   @Override
+                   public Result createOutput(final String namespaceUri, final String suggestedFileName) {
+                       final StreamResult result = new StreamResult(new CharArrayWriter());
+                       String systemId = "xsd" + (counter++) + ".xsd";
+                       result.setSystemId(systemId);
+                       results.add(result);
+                       return result;
+                   }
+               });
+
+               if (grammars != null) {
+                   Iterator<Include> iter = grammars.getInclude().iterator();
+                   while (iter.hasNext()) {
+                       for (Doc doc : iter.next().getDoc()) {
+                           if ("Generated".equals(doc.getTitle())) ;
```

Comment:
@liweinan This if statement is useless because of the semicolon at the end of line. Following line is executed every time. Is this a bug? 
cc @asoldano @marekkopecky

Reply:
yes it is.  But this part of code is redundant by current implementation, so it doesn't cause tests to fail. Let me create a JIRA issue to track this bug: https://issues.jboss.org/browse/RESTEASY-2058. Thanks for checking!